src/dotnet/commands/dotnet-buildserver/shutdown/BuildServerShutdownCommand.cs

```diff
@@ -14,130 +14,161 @@ namespace Microsoft.DotNet.Tools.BuildServer.Shutdown
+            {
+                _enumerationFlags |= ServerEnumerationFlags.MSBuild;
+            }
 
-                if (razor || all)
-                {
-                    enabledManagers.Add(new RazorServerManager());
-                }
+            if (vbcscompiler || all)
+            {
+                _enumerationFlags |= ServerEnumerationFlags.VBCSCompiler;
+            }
 
-                managers = enabledManagers;
+            if (razor || all)
+            {
+                _enumerationFlags |= ServerEnumerationFlags.Razor;
             }
 
-            Managers = managers;
+            _serverProvider = serverProvider ?? new BuildServerProvider();
             _useOrderedWait = useOrderedWait;
             _reporter = reporter ?? Reporter.Output;
             _errorReporter = reporter ?? Reporter.Error;
         }
 
-        public IEnumerable<IBuildServerManager> Managers { get; }
-
         public override int Execute()
         {
-            bool success = true;
-
             var tasks = StartShutdown();
 
+            if (tasks.Count == 0)
+            {
+                _reporter.WriteLine(LocalizableStrings.NoServersToShutdown.Green());
+                return 0;
+            }
+
+            bool success = true;
             while (tasks.Count > 0)
             {
                 var index = WaitForResult(tasks.Select(t => t.Item2).ToArray());
-                var (manager, task) = tasks[index];
+                var (server, task) = tasks[index];
 
-                success &= HandleResult(manager, task.Result);
+                if (task.IsFaulted)
+                {
+                    success = false;
+                    WriteFailureMessage(server, task.Exception);
+                }
+                else
+                {
+                    WriteSuccessMessage(server);
+                }
 
                 tasks.RemoveAt(index);
             }
 
             return success ? 0 : 1;
         }
 
-        private List<(IBuildServerManager, Task<Result>)> StartShutdown()
+        private List<(IBuildServer, Task)> StartShutdown()
         {
-            var tasks = new List<(IBuildServerManager, Task<Result>)>();
-            foreach (var manager in Managers)
+            var tasks = new List<(IBuildServer, Task)>();
+            foreach (var server in _serverProvider.EnumerateBuildServers(_enumerationFlags))
             {
-                _reporter.WriteLine(string.Format(LocalizableStrings.ShuttingDownServer, manager.ServerName));
-                tasks.Add((manager, manager.ShutdownServerAsync()));
+                WriteShutdownMessage(server);
+                tasks.Add((server, Task.Run(() => server.Shutdown())));
             }
+
             return tasks;
         }
 
         private int WaitForResult(Task[] tasks)
         {
             if (_useOrderedWait)
             {
-                tasks[0].Wait();
-                return 0;
+                return Task.WaitAny(tasks.First());
```

Comment:
Is this semantically different than what was here before.  I think it's the same, except if tasks is empty in which case IndexOutOfRangeException becomes InvalidOperationException? Am I misreading? Is this just a style preference? If it is a style preference, I don't have an opinion on it, but I just want to understand.

Reply:
It is, because `Task.Wait` raises the exception that was thrown for a faulting task.

Reply:
I see!